Mathlib/Data/Multiset/DershowitzManna.lean

```diff
@@ -0,0 +1,294 @@
+  refine ⟨fun y hy ↦ ?_, fun y hy ↦ ?_⟩
+  · obtain ⟨z, hz, hyz⟩ := hYZ₁ y hy
+    by_cases z_in : z ∈ Y₂
+    · obtain ⟨w, hw, hzw⟩ := hYZ₂ z z_in
+      exact ⟨w, .inl hw, hyz.trans hzw⟩
+    · exact ⟨z, .inr <| by rwa [mem_sub, count_eq_zero_of_not_mem z_in, count_pos], hyz⟩
+  · obtain ⟨z, hz, hyz⟩ := hYZ₂ y <| mem_of_le (Multiset.sub_le_self ..) hy
+    exact ⟨z, .inl hz, hyz⟩
+
+/-- A special case of `IsDershowitzMannaLT`. The transitive closure of it is used to define
+an equivalent (proved later) version of the ordering. -/
+private def OneStep (M N : Multiset α) : Prop :=
+  ∃ X Y a,
+      M = X + Y
+    ∧ N = X + {a}
+    ∧ ∀ y ∈ Y, y < a
+
+private lemma isDershowitzMannaLT_of_oneStep : OneStep M N → IsDershowitzMannaLT M N := by
+  rintro ⟨X, Y, a, M_def, N_def, ys_lt_a⟩
+  use X, Y, {a}, by simp, M_def, N_def
+  · simpa
+
+private lemma isDershowitzMannaLT_singleton_insert (h : OneStep N (a ::ₘ M)) :
+    ∃ M', N = a ::ₘ M' ∧ OneStep M' M ∨ N = M + M' ∧ ∀ x ∈ M', x < a := by
+  classical
+  rcases h with ⟨X, Y, a0, rfl, h0, h2⟩
+  by_cases hyp : a = a0
+  · exists Y; right; apply And.intro
+    · rw [add_left_inj]
+      rw [add_comm, singleton_add] at h0
+      simp_all
+    · simp_all
+  exists (Y + (M - {a0}))
+  left
+  constructor
+  · have : X = (M - {a0} + {a}) := by
+      rw [add_comm, singleton_add] at *
+      ext b
+      simp only [ext, count_cons] at h0
+      by_cases h : b = a
+      · have := h0 b
+        simp_all
+      have := h0 b
+      aesop
+    subst this
+    rw [add_comm]
+    nth_rewrite 2 [add_comm]
+    rw [singleton_add, add_cons]
+  · unfold OneStep
+    refine ⟨M - {a0}, Y, a0, ?_, ?_, h2⟩
+    · change Y + (M - {a0}) = (M - {a0}) + Y
+      rw [add_comm]
+    · change M = M - {a0} + {a0}
+      have this0 : M = X + {a0} - {a} := by
+        rw [← h0, sub_singleton, erase_cons_head]
+      have a0M : a0 ∈ M := by
+        rw [this0, sub_singleton, mem_erase_of_ne]
+        rw [mem_add, mem_singleton]
+        · apply Or.inr
+          rfl
+        · exact fun h ↦ hyp (Eq.symm h)
+      rw [add_comm]
+      simp_all
+
+private lemma acc_cons (a : α) (M0 : Multiset α)
+    (_ : ∀ b M , LT.lt b a → Acc OneStep M → Acc OneStep (b ::ₘ M))
+    (_ : Acc OneStep M0)
+    (_ : ∀ M, OneStep M M0 → Acc OneStep (a ::ₘ M)) :
+    Acc OneStep (a ::ₘ M0) := by
+  constructor
+  intros N N_lt
+  rcases (isDershowitzMannaLT_singleton_insert N_lt) with ⟨x, H, h0⟩
+  case h.intro.inr h =>
+    rcases h with ⟨H, h0⟩
+    rw [H]
+    clear H
+    induction x using Multiset.induction with
+    | empty =>
+      simpa
+    | cons h =>
+      simp_all
+  case h.intro.inl.intro =>
+    simp_all
+
+private lemma acc_cons_of_acc (a : α)
+    (H : ∀ b, ∀ M, b < a → Acc OneStep M → Acc OneStep (b ::ₘ M)) :
+    ∀ M, Acc OneStep M → Acc OneStep (a ::ₘ M) := by
+  intros M h0
+  induction h0 with
+  | intro x wfH wfh2 =>
+    apply acc_cons
+    · simpa
+    · constructor; simpa
+    · simpa
```

Comment:
```suggestion
    · assumption
    · constructor; assumption
    · assumption
```
`simpa` is (kinda) `simp` followed by `assumption`, if `assumption` alone works then it is better. This might apply to other places too.